pkg/controller/garbagecollector/garbage_collector.go

```diff
@@ -0,0 +1,501 @@
+	schema.GroupVersionResource{Group: "authorization.k8s.io", Version: "v1beta1", Resource: "localsubjectaccessreviews"}: {},
+}
+
+func NewGarbageCollector(metaOnlyClientPool dynamic.ClientPool, clientPool dynamic.ClientPool, mapper meta.RESTMapper, deletableResources map[schema.GroupVersionResource]struct{}) (*GarbageCollector, error) {
+	gc := &GarbageCollector{
+		metaOnlyClientPool:               metaOnlyClientPool,
+		clientPool:                       clientPool,
+		restMapper:                       mapper,
+		clock:                            clock.RealClock{},
+		dirtyQueue:                       workqueue.NewTimedWorkQueue(),
+		orphanQueue:                      workqueue.NewTimedWorkQueue(),
+		registeredRateLimiter:            NewRegisteredRateLimiter(deletableResources),
+		registeredRateLimiterForMonitors: NewRegisteredRateLimiter(deletableResources),
+		absentOwnerCache:                 NewUIDCache(500),
+	}
+	gc.propagator = &Propagator{
+		eventQueue: workqueue.NewTimedWorkQueue(),
+		uidToNode: &concurrentUIDToNode{
+			RWMutex:   &sync.RWMutex{},
+			uidToNode: make(map[types.UID]*node),
+		},
+		gc: gc,
+	}
+	for resource := range deletableResources {
+		if _, ok := ignoredResources[resource]; ok {
+			glog.V(6).Infof("ignore resource %#v", resource)
+			continue
+		}
+		kind, err := gc.restMapper.KindFor(resource)
+		if err != nil {
+			return nil, err
+		}
+		monitor, err := gc.monitorFor(resource, kind)
+		if err != nil {
+			return nil, err
+		}
+		gc.monitors = append(gc.monitors, monitor)
+	}
+	return gc, nil
+}
+
+func (gc *GarbageCollector) Run(workers int, stopCh <-chan struct{}) {
+	defer gc.dirtyQueue.ShutDown()
+	defer gc.orphanQueue.ShutDown()
+	defer gc.propagator.eventQueue.ShutDown()
+
+	glog.Infof("Garbage Collector: Initializing")
+	for _, monitor := range gc.monitors {
+		go monitor.controller.Run(stopCh)
+	}
+
+	var syncs []cache.InformerSynced
+	for _, monitor := range gc.monitors {
+		syncs = syncs.append(monitor.controller.HasSynced())
+	}
+	if !cache.WaitForCacheSync(stopCh, syncs...) {
+		return
+	}
+	glog.Infof("Garbage Collector: All monitored resources synced. Proceeding to collect garbage")
+
+	// worker
+	go wait.Until(gc.propagator.processEvent, 0, stopCh)
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(gc.worker, 0, stopCh)
+		go wait.Until(gc.orphanFinalizer, 0, stopCh)
+	}
+	Register()
+	<-stopCh
+	glog.Infof("Garbage Collector: Shutting down")
+}
+
+func (gc *GarbageCollector) worker() {
+	timedItem, quit := gc.dirtyQueue.Get()
+	if quit {
+		return
+	}
+	defer gc.dirtyQueue.Done(timedItem)
+	err := gc.processItem(timedItem.Object.(*node))
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Error syncing item %#v: %v", timedItem.Object, err))
+		// retry if garbage collection of an object failed.
+		gc.dirtyQueue.Add(timedItem)
+		return
+	}
+	DirtyProcessingLatency.Observe(sinceInMicroseconds(gc.clock, timedItem.StartTime))
```

Comment:
If I dequeue an object, process it, but an error occurs so I enqueue it once more, the normal work queue doesn't know it's the same object, right? I can try to expand the normal work queue to support such case.